src/main/java/net/neoforged/neoforge/event/ValidityEvent.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.event;
+
+import java.util.Optional;
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.entity.vehicle.ContainerEntity;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.neoforged.neoforge.common.NeoForge;
+import net.neoforged.neoforge.event.entity.player.PlayerEvent;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Event to check if a menu can stay open or if the player has moved too far away from it.
+ * <p>
+ * This is intended for mods that allow a player remote access to blocks and their menus,
+ * but it also can be used to forcibly kick them out of menus.
+ * <p>
+ * You can query the result of the original distance check with {@link #getOriginallyValid()}
+ * and set your override with {@link #setStillValid(boolean)}.
+ * <p>
+ * This base class is never posted (but it can be subscribed to, as it is not abstract),
+ * only one of its subclasses:
+ * <ul>
+ * <li>{@link ValidityEvent.Menu} for menus that have a {@link BlockPos} and may or may not be container-based,
+ * <li>{@link ValidityEvent.ContainerEntityMenu} for {@link ContainerEntity}-based menus, or
+ * <li>{@link ValidityEvent.EntityMenu} for menus attached to an {@link Entity}, or
+ * <li>{@link ValidityEvent.BlockEntityMenu} for the checks done by signs which don't use menus but a Screen
+ * and modded menu-like Screens that opt-in to use this.
+ * </ul>
+ * This event is fired on the {@link NeoForge#EVENT_BUS}.
+ */
+public class ValidityEvent extends PlayerEvent {
+    private final double distance;
+    private final boolean originallyValid;
+    private Optional<Boolean> stillValid = Optional.empty();
+
+    ValidityEvent(Player player, double distance, boolean originallyValid) {
+        super(player);
+        this.distance = distance;
+        this.originallyValid = originallyValid;
+    }
+
+    public double getDistance() {
+        return distance;
+    }
+
+    public boolean getOriginallyValid() {
+        return originallyValid;
+    }
+
+    public void setStillValid(boolean result) {
+        this.stillValid = Optional.of(result);
+    }
+
+    public boolean getStillValid() {
+        return stillValid.orElse(originallyValid);
+    }
+
+    public static class BlockBased extends ValidityEvent {
```

Comment:
Why is this not abstract? I'd also prefer if you made the subclasses sealed instead of package-private ctors.

Reply:
see above